src/analyzer.ts

```diff
@@ -73,45 +113,88 @@ export class Analyzer {
 
     // Note we can't use Promise.all here, because new promises can be added to
     // the promises array as long as any promise is pending.
-    const errors = [];
-    while (this.#placeholderUpgradePromises.length > 0) {
-      const result = await this.#placeholderUpgradePromises.shift();
-      if (result?.ok === false) {
-        errors.push(result.error);
+    await this.#waitForanalysisToComplete();
+    {
+      const errors = await this.getDiagnostics();
+      if (errors.size > 0) {
+        return {ok: false, error: [...errors]};
       }
     }
-    if (errors.length > 0) {
-      return {ok: false, error: errors};
-    }
 
     // We can safely assume all placeholders have now been upgraded to full
     // configs.
-    const rootConfig = rootPlaceholder as ScriptConfig;
+    const rootConfig = rootPlaceholder.placeholder;
+    if (rootConfig.state === 'unvalidated') {
+      throw new Error(
```

Comment:
Just checking, is the rule that we `throw` if it's an internal implementation error, because that's more convenient and in some cases can keep noise out of the type system?

Reply:
Yeah exactly. An internal error should never happen, but we can't rule it out with the type system, so just in case we throw an informative error.

Reply:
I guess we can think of `throw` like you would think of `panic` in Go or Rust.